### PR TITLE
feature: differentiate between achievement and event relation

### PR DESCRIPTION
**Pull request by dhenkel92**
This PR is decoupling the achievement relation from the event relation.
The idea is that an event is bound to a specific appointment, so that we can relate it to a bucket later on.

### CODE REVIEW

File: common/achievement/evaluate.ts [hunk 1]
```diff
@@ -95,6 +96,8 @@ async function _evaluateAchievement(
         if (skipAchievementBucketsAfter) {
             bucketEvents = removeBucketsAfter(skipAchievementBucketsAfter, bucketEvents, true);
         }
+        console.log('bucketEvents', JSON.stringify(bucketEvents, null, 2));
+        console.log('events', JSON.stringify(eventsForMetric, null, 2));
```
Comment (dhenkel92) [inline common/achievement/evaluate.ts:100]: TODO: remove

File: common/achievement/evaluate.ts [hunk 2]
```diff
@@ -27,7 +27,8 @@ async function _evaluateAchievement(
         where: {
             userId,
             metric: { in: metrics },
-            AND: relation ? { relation: { startsWith: relation } } : {},
+            // TODO:
```
Comment (dhenkel92) [inline common/achievement/evaluate.ts:30]: TODO

File: common/achievement/util.ts [hunk 3]
```diff
@@ -217,3 +208,61 @@ export function checkIfAchievementIsGlobal(template: achievement_template) {
...
+
+    const [type, id] = parseRelation(event.context.relation);
+    if (type === EventRelationType.Match || type === EventRelationType.Subcourse) {
+        return event.context.relation;
+    }
+
+    if (type === EventRelationType.Appointment) {
+        const appointment = await prisma.lecture.findUnique({ where: { id } });
+        if (!appointment) {
+            // TODO: throw an error?
```
Comment (dhenkel92) [inline common/achievement/util.ts:254]: TODO:

File: common/achievement/util.ts [hunk 4]
```diff
@@ -217,3 +208,61 @@ export function checkIfAchievementIsGlobal(template: achievement_template) {
...
+
+        if (appointment.subcourseId) {
+            return createRelation(EventRelationType.Subcourse, appointment.subcourseId);
+        }
+        if (appointment.matchId) {
+            return createRelation(EventRelationType.Match, appointment.matchId);
+        }
+    }
+
+    // TODO: throw an error?
```
Comment (dhenkel92) [inline common/achievement/util.ts:266]: TODO:

File: common/achievement/evaluate.ts [hunk 5]
```diff
@@ -27,7 +27,6 @@ async function _evaluateAchievement(
         where: {
             userId,
             metric: { in: metrics },
-            AND: relation ? { relation: { startsWith: relation } } : {},
```
Comment (dhenkel92) [inline common/achievement/evaluate.ts:30]: Unfortunately, I cannot think of a better way right now, but the impact should not be very limited.

**Review by Jonasdoubleyou — APPROVED** [review]

**Pull request merged**